src/manifold/src/smoothing.cpp

```diff
@@ -421,73 +610,258 @@ void Manifold::Impl::CreateTangents(
 }
 
 /**
- * Split each edge into n pieces and sub-triangulate each triangle accordingly.
- * This function doesn't run Finish(), as that is expensive and it'll need to be
- * run after the new vertices have moved, which is a likely scenario after
- * refinement (smoothing).
+ * Split each edge into n pieces as defined by calling the edgeDivisions
+ * function, and sub-triangulate each triangle accordingly. This function
+ * doesn't run Finish(), as that is expensive and it'll need to be run after the
+ * new vertices have moved, which is a likely scenario after refinement
+ * (smoothing).
  */
-Vec<Barycentric> Manifold::Impl::Subdivide(int n) {
-  ZoneScoped;
-  if (n < 2) return Vec<Barycentric>();
-  faceNormal_.resize(0);
-  vertNormal_.resize(0);
-  const int numVert = NumVert();
-  const int numEdge = NumEdge();
-  const int numTri = NumTri();
-  // Append new verts
-  const int vertsPerEdge = n - 1;
-  const int triVertStart = numVert + numEdge * vertsPerEdge;
-  vertPos_.resize(triVertStart + numTri * VertsPerTri(n - 2));
-  Vec<Barycentric> vertBary(vertPos_.size());
-  FillRetainedVerts(vertBary, halfedge_);
-
-  MeshRelationD oldMeshRelation = std::move(meshRelation_);
-  meshRelation_.triRef.resize(n * n * numTri);
-  meshRelation_.originalID = oldMeshRelation.originalID;
-
+Vec<Barycentric> Manifold::Impl::Subdivide(
+    std::function<int(glm::vec3)> edgeDivisions) {
```

Comment:
Wonder if the division function can be more powerful. For example instead of accepting the edge as a vector, maybe we can get the edge as an index in the original mesh, so we can implement something like remeshing that produce delaunay triangulation? (although iirc those are usually iterative algorithms)

Reply:
I think the more powerful thing I want to do is also use the edge's tangent vectors - that's my plan for `RefineToPrecision`. Delaunay is great, but you still need to place the interior verts, plus it's geometric, so no chance to cache like this one.